flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/AbstractTaskManagerFileHandler.java

```diff
@@ -0,0 +1,236 @@
+
+	private final LoadingCache<ResourceID, CompletableFuture<TransientBlobKey>> fileBlobKeys;
+
+	protected AbstractTaskManagerFileHandler(
+			@Nonnull CompletableFuture<String> localAddressFuture,
+			@Nonnull GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+			@Nonnull Time timeout,
+			@Nonnull Map<String, String> responseHeaders,
+			@Nonnull UntypedResponseMessageHeaders<EmptyRequestBody, M> untypedResponseMessageHeaders,
+			@Nonnull GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever,
+			@Nonnull TransientBlobService transientBlobService,
+			@Nonnull Time cacheEntryDuration) {
+		super(localAddressFuture, leaderRetriever, timeout, responseHeaders, untypedResponseMessageHeaders);
+
+		this.resourceManagerGatewayRetriever = Preconditions.checkNotNull(resourceManagerGatewayRetriever);
+
+		this.transientBlobService = Preconditions.checkNotNull(transientBlobService);
+
+		this.fileBlobKeys = CacheBuilder
+			.newBuilder()
+			.expireAfterWrite(cacheEntryDuration.toMilliseconds(), TimeUnit.MILLISECONDS)
+			.removalListener(this::removeBlob)
+			.build(
+				new CacheLoader<ResourceID, CompletableFuture<TransientBlobKey>>() {
+					@Override
+					public CompletableFuture<TransientBlobKey> load(ResourceID resourceId) throws Exception {
+						return loadTaskManagerFile(resourceId);
+					}
+			});
+	}
+
+	@Override
+	protected void respondToRequest(ChannelHandlerContext ctx, HttpRequest httpRequest, HandlerRequest<EmptyRequestBody, M> handlerRequest, RestfulGateway gateway) throws RestHandlerException {
+		final ResourceID taskManagerId = handlerRequest.getPathParameter(TaskManagerIdPathParameter.class);
+
+		final CompletableFuture<TransientBlobKey> blobKeyFuture;
+		try {
+			blobKeyFuture = fileBlobKeys.get(taskManagerId);
+		} catch (ExecutionException e) {
+			final Throwable cause = ExceptionUtils.stripExecutionException(e);
+			if (cause instanceof RestHandlerException) {
+				throw (RestHandlerException) cause;
+			} else {
+				throw new RestHandlerException("Could not retrieve file blob key future.", HttpResponseStatus.INTERNAL_SERVER_ERROR, e);
+			}
+		}
+
+		final CompletableFuture<Void> resultFuture = blobKeyFuture.thenAcceptAsync(
+			(TransientBlobKey blobKey) -> {
+				final File file;
+				try {
+					file = transientBlobService.getFile(blobKey);
+				} catch (IOException e) {
+					throw new CompletionException(new FlinkException("Could not retrieve file from transient blob store.", e));
+				}
+
+				try {
+					transferFile(
+						ctx,
+						file,
+						httpRequest);
+				} catch (FlinkException e) {
+					throw new CompletionException(new FlinkException("Could not transfer file to client.", e));
+				}
+			},
+			ctx.executor());
+
+		resultFuture.whenComplete(
+			(Void ignored, Throwable throwable) -> {
+				if (throwable != null) {
+					log.debug("Failed to transfer file from TaskExecutor {}.", taskManagerId, throwable);
+					fileBlobKeys.invalidate(taskManagerId);
```

Comment:
It's not enough to invalidate the cache, you must also send an error back to the client, or the connection will not be closed.

This will block forever if the TM is not registered:
```
curl -v http://localhost:9067/taskmanagers/daecac46c3f0f13b945fd2bb94438204/log
```

Reply:
This is true. Will add an error response here.